awslabs/aws-lambda-powertools-typescript
# refactor(tracer): log warning instead of throwing when segment is not found

## Description of your changes

As reported in #1356, the current behavior of the Tracer and corresponding middleware is fairly unforgiving when it comes to handling missing segments. Prior to this PR, when calling the `getSegment` method, if a segment is not found, an error would be thrown.

This can happen either when users try to retrieve the current segment outside of the context of a function invocation or when the trace data is malformed and the X-Ray SDK is not able to derive the current segment.

This behavior, if not handled properly, can cause a function invocation to fail, which might not be desirable given that for many customers tracing is considered best-effort.

This PR:
- aligns the Tracer utility with the X-Ray SDK so that if a segment is not found, a warning is logged instead of throwing
- improves the handling of missing segments in the `captureLambdaHandler` middleware in all stages to be more resilient
- bumps the `aws-xray-sdk-node` dependency to the latest version
- refactor and extracts some of the internal logic around current segment retrieval from the `Tracer` class to the `ProviderService` one

This PR closes #1356.

### How to verify this change

Check the existing tests as well as the newly added ones in the checks under this repo. Check also the results of [this successful integration test](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/4407673701).

### Related issues, RFCs



**Issue number:** #1356

## Checklist


### Breaking change checklist

***Is it a breaking change?:*** NO


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.
